network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+    let full_node_infos = full_node_set.payload().to_vec();
+
+    let mut updates = Vec::new();
+
+    // Collect the set of updated addresses.
+    let address_map = full_node_infos
+        .iter()
+        .filter(|full_node| {
+            match NetworkAddress::try_from(&full_node.config().full_node_network_address) {
+                Ok(_) => true,
+                Err(e) => {
+                    warn!(
+                        "Cannot parse validator network address from validator info: {}",
+                        e
+                    );
+                    false
+                }
+            }
+        })
+        .map(|full_node| {
+            (
+                *full_node.account_address(),
+                vec![NetworkAddress::try_from(
+                    &full_node.config().full_node_network_address.clone(),
+                )
+                .expect("Already filtered, FATAL")],
+            )
+        })
+        .collect();
+
+    let update_address_req =
+        ConnectivityRequest::UpdateAddresses(DiscoverySource::OnChain, address_map);
+    updates.push(update_address_req);
+
+    // Collect the EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        full_node_infos
+            .into_iter()
+            .map(|full_node| {
+                (
+                    *full_node.account_address(),
+                    NetworkPublicKeys {
+                        identity_public_key: full_node
+                            .config()
+                            .full_node_network_identity_public_key,
+                        // TODO: this overload of the validator key onto the full_node network is
+                        // confusing at best and misleading at worst.
+                        signing_public_key: full_node
+                            .config()
+                            .validator_network_signing_public_key
+                            .clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+
+    updates
+}
+
+impl ConfigurationChangeListener {
+    /// Creates a new ConfigurationListener
+    pub fn new(conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>, role: RoleType) -> Self {
+        Self {
+            conn_mgr_reqs_tx,
+            role,
+        }
+    }
+
+    /// Processes a received OnChainConfigPayload.  Depending on role (Validator or FullNode), parses
+    /// the appropriate configuration changes and passes it to the ConnectionManager channel.
+    async fn process_payload(&mut self, payload: OnChainConfigPayload) {
+        let node_set: ValidatorSet = payload
+            .get()
+            .expect("failed to get ValidatorSet from payload");
+
+        let updates = match self.role {
+            RoleType::Validator => extract_validator_updates(node_set),
+            RoleType::FullNode => extract_full_node_updates(node_set),
```

Comment:
these two functions still look like the same except the address it's extracted, probably can be simplified with a lambda just to get the address,
```
extract_updates(validator_set: ValidatorSet, address: Fn(Config) -> NetworkAddress)
```

Reply:
address and public_key;  but yes.